Lua task preempts long running scripts instead of killing them

jfrickmann opened the pull request:
Hand ported from https://github.com/opentx/opentx/pull/8194.
It will not merge automatically, and I think that it is due to changes in libopenui, which I have not touched. Could a Git-fu master please give me a little help here?

raphaelcoeffic commented:
> Hand ported from [opentx/opentx#8194](https://github.com/opentx/opentx/pull/8194).
> It will not merge automatically, and I think that it is due to changes in libopenui, which I have not touched. Could a Git-fu master please give me a little help here?

Can you please invite me to your repo, so I can change it for you? Otherwise I have no access to your branch.

jfrickmann commented:
OK @raphaelcoeffic I have hereby invited you to my repo.

raphaelcoeffic commented:
> OK @raphaelcoeffic I have hereby invited you to my repo.

Thank you Sir! I just rebased on top of current 2.4 for clarity. The issue with `libopenui` is solved as well.

raphaelcoeffic commented:
Oooops! It looks like I screwed some changes on the way....

```
/__w/edgetx/edgetx/radio/src/lua/interface.cpp: In function 'void displayLuaError(const char*)':
/__w/edgetx/edgetx/radio/src/lua/interface.cpp:691:24: error: 'WARNING_LINE_X' was not declared in this scope
  691 |       lcdDrawSizedText(WARNING_LINE_X, WARNING_LINE_Y+FH+3, lua_warning_info, split-lua_warning_info, SMLSIZE);
      |                        ^~~~~~~~~~~~~~
/__w/edgetx/edgetx/radio/src/lua/interface.cpp:691:40: error: 'WARNING_LINE_Y' was not declared in this scope
  691 |       lcdDrawSizedText(WARNING_LINE_X, WARNING_LINE_Y+FH+3, lua_warning_info, split-lua_warning_info, SMLSIZE);
      |                                        ^~~~~~~~~~~~~~
/__w/edgetx/edgetx/radio/src/lua/interface.cpp:691:103: error: 'SMLSIZE' was not declared in this scope
  691 |       lcdDrawSizedText(WARNING_LINE_X, WARNING_LINE_Y+FH+3, lua_warning_info, split-lua_warning_info, SMLSIZE);
      |                                                                                                       ^~~~~~~
```

They were probably not in separate commits, but in the merge commits from 2.4, right?

jfrickmann commented:
Most likely I screwed up something when I edited the file. I can see that WARNING_LINE_X is defined for 128x64 and 212x64 screens only. I have to go to work now, but please let me take a look at it tonight and see if I can fix it.

raphaelcoeffic commented:
> Most likely I screwed up something when I edited the file. I can see that WARNING_LINE_X is defined for 128x64 and 212x64 screens only. I have to go to work now, but please let me take a look at it tonight and see if I can fix it.

It compiles fine now. But I have not tried anything, which you can do now ;-) I'm not quite sure I did it correctly for the widgets, please have a look. There is a `luaTask` call there, and I'm really not sure this is intentional this way.

jfrickmann commented:
I need a little help here...  While it seems to work fine for X9D+, it appears to enter an infinite loop when I test the FrSky SBEC standalone script for a TX16S build in the simulator. displayLuaError(...) in lua/interface.cpp calls POPUP_WARNING(...), which never returns. Apparently, while (popup->getParent()) { ... } keeps looping here, after the popup has been dismissed.

jfrickmann commented:
I think that I figured it out. Because ModalWindow takes over running the MainWindow, StandaloneLuaWindow::runLua(...) has to be re-entrant.

jfrickmann commented:
I have now built the firmware with this PR and successfully tested it on a TX12, TX16S and a X9D+. No flight testing yet, but the tool, standalone and telemetry scripts all run w/o problems. So I feel that it is ready to be merged now.

pfeerick commented:
What sort of stuff do I need to do to test this? Just run some lua scripts via tools and see if they behave? Any particular ones that are best suited?

jfrickmann commented:
Yes, the first level would be to just run anything that you have to make sure that I did not break something. Please notice that I should not have changed anything for widgets, but it would probably still be a good idea to verify that. I have tested the Tools scripts and my own SoarOTX scripts. I used the following little standalone script to test preemption:

```
local i = 0

local function run(event)
  lcd.clear()
  local t = getTime() + 100
  repeat until getTime() >= t
  i = i + 1
  lcd.drawNumber(LCD_W / 2, LCD_H / 2, i, DBLSIZE)
  return 0
end

return { run = run }
```
It tests two aspects: 1) that it can busy loop for 1 sec. 2) that screen updating works, i.e. the screen does not flicker.

The original motivation for this PR is the [graphing script](https://github.com/jfrickmann/SoarOTX/blob/master/SD-CARD/SCRIPTS/TELEMETRY/JF/GRREAD.lua) in my SoarOTX suite. Reading data from the log file is painfully slow, because I had to reduce the number of lines read per cycle to only 4, to avoid getting killed by the CPU limit. I suspect that it does a busy wait for I/O and that sometimes racks up the instruction count. To test preemption, I increased READ_MAX (line 32) from 4 to 9999 and subjectively, it reads the log files much faster now.

I originally made this PR over the Xmas holiday to OTX 2.3, and I have been flying two DLGs with that build on my radio since New Year. So the main purpose here is to verify that I did not screw up something when I ported the PR to ETX.

lshems commented:
I will test it as well, without the Lua build option enables.lol.

pfeerick commented:
I just had a run through of this PR (and enabled GitHub builds as I didn't realise before it wasn't auto-building yet) ... ran through things like the betaflight and inav scripts, some other dev scripts, multiprotocol module scripts, your own test script, turned on a bunch of widgets... and I haven't seen any harmful side effects... scripts that worked before are still working and nothing seems to be crashing. Hm... that reminds me... have to go download the ExpressLRS lua script now...

jfrickmann commented:
Great! I had two DLG flying sessions during the weekend with this build running my SoarOTX scripts and total flying time around 3 hrs without encountering any problems.

jfrickmann commented:
I have added a commit that passes key events to full screen widgets. Instead of trying to find the Zone Lua table and modify it, we can just use LCD_W and LCD_H to get the screen dimensions in fullscreen. We can test if event == nil to see if we are in fullscreen mode. Here is a little test widget:
```
local defaultOptions = {
  { "ControlX", SOURCE, 1 },
  { "ScrollZ", BOOL, 1 }, -- BOOL is actually not a boolean, but toggles between 0 and 1
  { "StepZ", VALUE, 1, 0, 10},
  { "COLOR", COLOR, RED },
}

local function create(zone, options)
  return { zone=zone, options=options }
end

local function update(widget, options)
  widget.options = options
  lcd.setColor( CUSTOM_COLOR, widget.options.COLOR )
end

local function background(widget)
end

local function refresh(widget, event)
  if event == nil then
    x = widget.zone.x
    y = widget.zone.y
    w = widget.zone.w
    h = widget.zone.h
    lcd.drawText(10, 10, "Small");
  else
    x = 0
    y = 0
    w = LCD_W
    h = LCD_H
    lcd.drawText(10, 10, "Full");
  end
  
  for i = 0, 5 do
    lcd.drawRectangle(x+ i, y + i, w - 2 * i, h - 2 * i)
  end
end

return { 
  name = "JF-Test", 
  options = defaultOptions, 
  create = create, 
  update = update, 
  refresh = refresh, 
  background = background
}
```

pfeerick commented:
I was about to comment on that...  nice! Is it possible (or necessary?) to
have it not pass a long return (I am presuming here that it should not be
passed so that it remains reserved for getting out of full screen... or am
I confusing  my full screen luas?)?



On Wed, 7 Jul 2021, 12:14 pm Jesper Frickmann, ***@***.***>
wrote:

> I have added a commit that passes key events to full screen widgets.
> Instead of trying to find the Zone Lua table and modify it, we can just use
> LCD_W and LCD_H to get the screen dimensions in fullscreen. We can test if
> event == nil to see if we are in fullscreen mode. Here is a little test
> widget:
>
> local defaultOptions = {
>   { "ControlX", SOURCE, 1 },
>   { "ScrollZ", BOOL, 1 }, -- BOOL is actually not a boolean, but toggles between 0 and 1
>   { "StepZ", VALUE, 1, 0, 10},
>   { "COLOR", COLOR, RED },
> }
>
> local function create(zone, options)
>   return { zone=zone, options=options }
> end
>
> local function update(widget, options)
>   widget.options = options
>   lcd.setColor( CUSTOM_COLOR, widget.options.COLOR )
> end
>
> local function background(widget)
> end
>
> local function refresh(widget, event)
>   if event == nil then
>     x = widget.zone.x
>     y = widget.zone.y
>     w = widget.zone.w
>     h = widget.zone.h
>     lcd.drawText(10, 10, "Small");
>   else
>     x = 0
>     y = 0
>     w = LCD_W
>     h = LCD_H
>     lcd.drawText(10, 10, "Full");
>   end
>
>   for i = 0, 5 do
>     lcd.drawRectangle(x+ i, y + i, w - 2 * i, h - 2 * i)
>   end
> end
>
> return {
>   name = "JF-Test",
>   options = defaultOptions,
>   create = create,
>   update = update,
>   refresh = refresh,
>   background = background
> }
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/EdgeTX/edgetx/pull/285#issuecomment-875216119>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABJ66KIXBBCVUCTV2VERK7TTWO2BPANCNFSM47AL75PA>
> .
>

jfrickmann commented:
I don't know if it matters, but I think that it gets filtered out in Widget::onEvent()

pfeerick commented:
Looking at that, yeah, I "learned" something today ;) It looks like it gets filtered out, so it's done properly already :)

lshems commented:
Super Jesper. This is going to be a deal maker for switching to edge.

I will probably switch completely to edge as soon as there is a formal.rekease. ditching official support for opentx, as it doesn't cater for the needs of the mainstream user anymore.

Might loose some friends,but I really believe the progress on edge and openness to go fourth on tough issues is refreshing and needed.

Thanks all again. Let's keep this running.

wimalopaan commented:
I tested this and it works as detecting fullscreen by event != nil.
But in fullscreen mode event ist always 0. I do not see any event codes coming in.

jfrickmann commented:
OK, I did not yet test that. Tonight I will see if I can override Widget::onEvent(event) to grab the event for LuaWidget.

raphaelcoeffic commented:
> OK, I did not yet test that. Tonight I will see if I can override Widget::onEvent(event) to grab the event for LuaWidget.

I can have a look this afternoon, if you want, as this is more or less my mess :-D

raphaelcoeffic commented:
> I tested this and it works as detecting fullscreen by event != nil.
> But in fullscreen mode event ist always 0. I do not see any event codes coming in.
> Tested in simulator24

Can you please post the script you are using to demonstrate that? That would save me a bit of time ;-)

jfrickmann commented:
OK, I have added a line that prints the last received event number. I would add an event member to LuaWidget, override with LuaWidget::onEvent(event) to save the event there if (fullscreen and EVT_KEY_LONG(KEY_EXIT) != event), and then pass the saved event to Lua in fullscreen.
```
local myEvent = 0

local defaultOptions = {
  { "ControlX", SOURCE, 1 },
  { "ScrollZ", BOOL, 1 }, -- BOOL is actually not a boolean, but toggles between 0 and 1
  { "StepZ", VALUE, 1, 0, 10},
  { "COLOR", COLOR, RED },
}

local function create(zone, options)
  return { zone=zone, options=options }
end

local function update(widget, options)
  widget.options = options
  lcd.setColor( CUSTOM_COLOR, widget.options.COLOR )
end

local function background(widget)
end

local function refresh(widget, event)
  if event == nil then
    x = widget.zone.x
    y = widget.zone.y
    w = widget.zone.w
    h = widget.zone.h
    lcd.drawText(10, 10, "Small");
    myEvent = 0
  else
    x = 0
    y = 0
    w = LCD_W
    h = LCD_H
    lcd.drawText(10, 10, "Full");
    if event ~= 0 then myEvent = event end
  end
  
  for i = 0, 5 do
    lcd.drawRectangle(x+ i, y + i, w - 2 * i, h - 2 * i)
  end

  lcd.drawNumber(10, 30, myEvent);
end

return { 
  name = "JF-Test", 
  options = defaultOptions, 
  create = create, 
  update = update, 
  refresh = refresh, 
  background = background
}
```

lshems commented:
Be aware that myEvent is now shared with all widget instances of this widget!

jfrickmann commented:
OK, I think that I got it working now by overriding onEvent in LuaWidget and adding an event member variable to store it before passing to Lua in refresh(). Please test :-) Here is the current test widget, with Guido's comment above fixed:
```
local defaultOptions = {
  { "COLOR", COLOR, RED },
}

local function create(zone, options)
  return { zone=zone, options=options, event = 0 }
end

local function update(widget, options)
  widget.options = options
  lcd.setColor( CUSTOM_COLOR, widget.options.COLOR )
end

local function background(widget)
end

local function refresh(widget, event)
  if event == nil then
    x = widget.zone.x
    y = widget.zone.y
    w = widget.zone.w
    h = widget.zone.h
    lcd.drawText(10, 10, "Small");
    widget.event = 0
  else
    x = 0
    y = 0
    w = LCD_W
    h = LCD_H
    lcd.drawText(10, 10, "Full");
    if event ~= 0 then
      print("-----> " .. event)
      widget.event = event
    end
  end
  
  for i = 0, 5 do
    lcd.drawRectangle(x+ i, y + i, w - 2 * i, h - 2 * i)
  end

  lcd.drawNumber(10, 30, widget.event);
end

return { 
  name = "JF-Test", 
  options = defaultOptions, 
  create = create, 
  update = update, 
  refresh = refresh, 
  background = background
}
```

wimalopaan commented:
Best! 
Works for me!!!

One small caveat: instead of adapting the zone dimensions by the lua script itself, the zon-table passed as an entry of the widget parameter to refresh() should contains the correct dimensions accordings to the widget-mode. 
This is in accordance to the other possibility to modify the zone-layout say from 1x1 to 2x4 where the zone dimensions are also passed in correct.

jfrickmann commented:
The system only passes zone dimensions to the widget in the call to create(…). It is up to the script writer to save the zone table in the widget data table. What if someone chooses not to do that, or calls it something else, e.g. “size”?
I am playing the Devil’s advocate here, and I think that we should do that when we design systems.  
Another way to look at it, is to think of full screen mode as breaking out of the zone. I think of it as a telemetry screen on X7 or X9D radios, and I would use a completely different layout in full screen vs. widget mode.
This is of course only my opinion, but I would like discuss before we start hacking code to find the zone table in Lua and edit it from system code during transitions to and from full screen mode.

lshems commented:
Agree on one time size passthrough on creation.

Keep it simple.

jfrickmann commented:
I thought a little more about it, and we could store a reference for the zone table before passing it to the create() function. That way, we can easily edit it when transitioning to/from full screen. I will try it out...

pfeerick commented:
Maybe I'm missing something here... isn't zone the size of the widget? Thus determined by the layout - i.e. 1x2, 2x2, 1x1, etc? Meaning you would want it to update on more than just create, but always be current, to account for the widget being resized (i.e. from 2x2 to fullscreen). And thus it would also be 'read-only' as far as the widget is concerned, because it can't override the layout?

wimalopaan commented:
Well, I think, an interface should be easily usable in the correct way, and only hard to be misused.
That said, if there is a zone entry in the parameter table to refresh(), this should be correct. If that could not be guaranteed, then the zone table should not be passed to refresh(). If so, the lua script has the reponsibility to request the zone-dimensions every time it needs to, e.g. with a seperate call to a function like zone() returning the actual zone-dimensions.

jfrickmann commented:
The zone table is only passed to the create() function. The script writer customarily stores this table in the widget data table, which is passed to the refresh() function. I agree that a function returning the current zone parameters would be a better solution, but it would also break the current API. And if we have to do that anyway, why not just accept that the screen size in fullscreen mode is given by LCD_W and LCD_H?

wimalopaan commented:
> The zone table is only passed to the create() function. The script writer customarily stores this table in the widget data table, which is passed to the refresh() function. 

Oh yes, I missed that fact. Thank you for pointing that out!

jfrickmann commented:
So, to recap the "key events to widgets in full screen mode":

- An event argument has been added to the refresh(...) function call.
- It is nil in widget mode and an integer in full screen mode. 0, if no keys were pressed in full screen mode.
- Hence, the Lua script can detect full screen mode by testing "event ~= nil".
- The "zone" table is created and passed to the create(...) function when it is called at the time of Lua widget creation.
- Normally, the script writer will store the zone table in the widget data table, and it is not being touched again by the system.
- The Lua constants LCD_W and LCD_H can be used to find the screen dimensions in full screen mode.

Is everyone onboard with this?

wimalopaan commented:
Would this approach also be possible for swipe-up/down/left/right events?

And possible tap events? Tap events would also require a coordinate-pair.

jfrickmann commented:
It probably would be possible for touch events too, albeit significantly more complicated than key events. If I understand it correctly, Raphael is still actively developing the touch interface in the GUI, so maybe we should wait for it to reach a somewhat mature level first. 
I do not see anything about tap events anywhere in the code. I see the basic interface in touch.h:
```
enum TouchEvent
{
    TE_NONE,
    TE_DOWN,
    TE_UP,
    TE_SLIDE,
    TE_SLIDE_END
};

struct TouchState
{
  unsigned char event;
  short x;
  short y;
  short startX;
  short startY;
  short deltaX;
  short deltaY;
  short lastDeltaX;
  short lastDeltaY;
};
```
And I see some onTouch functions implemented in libopengui in the Window class, which AFAIK is the base class for most other GUI elements:
```
bool Window::onTouchStart(coord_t x, coord_t y)
bool Window::onTouchEnd(coord_t x, coord_t y)
bool Window::onTouchSlide(coord_t x, coord_t y, coord_t startX, coord_t startY, coord_t slideX, coord_t slideY)

```

lshems commented:
Olliw has done some stuff for opentx. Perhaps have a look there?

raphaelcoeffic commented on radio/src/lua/interface.cpp:
```diff
@@ -144,41 +131,40 @@ void luaHook(lua_State * L, lua_Debug *ar)
 }
 
-int luaGetInputs(lua_State * L, ScriptInputsOutputs & sid)
+int luaGetInputs(lua_State * L2, ScriptInputsOutputs & sid)
```
@jfrickmann is there a particular reason for using that cryptic `L2` parameter name?

jfrickmann replied:
Yes indeed! To support preemption, Lua is called with lua_resume() instead of lua_pcall(), and L2 is the coroutine used for that. I have to use a coroutine instead of the main thread, because Lua has a "feature" that kills the thread forever if a script error happens during lua_resume(), and I can replace the coroutine but not the main thread. 
And in this function, I use both the coroutine and the main thread, because I preserve a string value by transferring it over to the main stack so that it does not get collected. The problem is that the string value otherwise would get dereferenced in Lua while an external reference is used. You can see the code from line 155:
```
        switch (field) {
          case 0:
            luaL_checktype(L2, -2, LUA_TNUMBER); // key is number
            luaL_checktype(L2, -1, LUA_TSTRING); // value is string
            lua_xmove(lsScripts, L, 1);          // To preserve the string value, move it to the main stack
            lua_pushnil(lsScripts);              // Keep the stack balanced
            lua_insert(L, -2);                   // Keep the coroutine at the top of the main stack
            si->name = lua_tostring(L, -2); // External reference to the string!
            break;
```

raphaelcoeffic replied:
Ok, so `L` is a global variable, and you changed the param name to `L2` to be able to use both at the same time, right?

raphaelcoeffic replied:
Maybe the function parameter should be more explicit, so that there is no possible confusion with the global variables?

jfrickmann replied:
Yes. And we can rename the function parameter if you like to make it more clear. No problem :-)
Do you want to do it, or shall I do it? And what should we call it?

raphaelcoeffic replied:
Your PR, your choice! I'm just here for critics :-D

jfrickmann replied:
Then I will use the opportunity to clean up some legacy confusion in the code :-) 
luaGetInputs and luaGetOutputs are only called from one place in the code, with L2 = lsScripts. So I will simply eliminate the L2 argument, and replace L2 with lsScripts. That should hopefully make things crystal clear!

jfrickmann replied:
Please take a look now, I have cleaned it up and also added directives to exclude unused Mixer/Custom/Model script code when LUA != YES.

wimalopaan commented:
Looks good, does work here.

jfrickmann commented:
Touch events have been added as well! Please see #422 for a little demo widget.

lshems commented:
Cab someone explain what I need to do to compile with this option in?

I want to test.

I have a github sync on the main branch, a working compile environment on Ubuntu.

Thanks.

jfrickmann commented:
You need to clone my for and compile from there: 
https://github.com/jfrickmann/edgetx

lshems commented:
Super. Thanks. Lot of things to learn.

pfeerick commented:
> You need to clone my fork and compile from there

Why can't use one of the GitHub autobuilds of this branch available under the "Checks" tab? These builds were working for me for testing this PR before.

lshems commented:
OK, that looks nice. So why do I bother setting up a build environment then?
Yes, I need to compile it myself for the simulator. And compiling only works on Linux, so the simulator will only work on linux as well (Ubuntu in my case.
But people that want to test a pull request on a physical radio can just get their firmware from the artifacts tab???

Why isn't that explained? I have spent litterally days on trying to find out how to test stuff. Nobody seems to be able to tell.

This is not an offense to nobody. It is just finding out that there are easy ways to test things.

lshems commented:
By the way, I really need to clone the fork from Jesper completely to be able to comile it. There is no other way to pull in his pull request, and not having to have a complete copy of everything where github knows exactly what is different?

I'm so confused about github. Sorry ....

jfrickmann commented:
> > You need to clone my fork and compile from there
> 
> Why can't use one of the GitHub autobuilds of this branch available under the "Checks" tab (Artifacts top right)? These builds were working for me for testing this PR before.

I did not know that you could do that either …

pfeerick commented:
It has been mentioned a few times, both on the Discord, and also on pull requests/issues before ;) Any pull request can get the 'standard' builds of firmware this way, either directly through github, or via the Flasher (this is what the `edgetx-firmware-merge` option serves up - it's just a matter of finding which commit belongs to which which PR). 

>By the way, I really need to clone the fork from Jesper completely to be able to comile it.

No, not so. There is a way to checkout a pull request. If you have the github cli client installed (`gh`, *not* `git`). you can just do `gh pr #pr-number`. It is [possible with `git` also](https://stackoverflow.com/a/30584951), but I've never bothered trying that way.

raphaelcoeffic commented:
@jfrickmann could you please rebase on top of `main` branch please? `2.4` will be left as a maintenance branch `2.4.x`, with `2.5` being next quarter's release.

jfrickmann commented:
> @jfrickmann could you please rebase on top of `main` branch please? `2.4` will be left as a maintenance branch `2.4.x`, with `2.5` being next quarter's release.

I will be happy to do so, I just don't know how. Sorry for my lack of git skills!

jfrickmann commented:
Just for me to learn - how did you do it?

raphaelcoeffic commented:
> > @jfrickmann could you please rebase on top of `main` branch please? `2.4` will be left as a maintenance branch `2.4.x`, with `2.5` being next quarter's release.
> 
> I will be happy to do so, I just don't know how. Sorry for my lack of git skills!

Done! Please note that you will have to be careful when checking out that code. You will need to "force pull" (or "hard reset" the head to the new commit)

raphaelcoeffic commented:
> Just for me to learn - how did you do it?

Part 1 is just "edit" the PR here on GH to have the base you want. Then:
```
git rebase origin/main
```

If it has any conflict, you need to "repair" them, and then go on with:
```
git rebase --continue
```

Regarding the issue with `VCENTERED`, where do you add it? It seems to fail compilation.

jfrickmann commented:
> Regarding the issue with `VCENTERED`, where do you add it? It seems to fail compilation.

lua/api_general.cpp line 1870. I would move it up in a COLORLCD block, e.g. line1864 under FONT(BOLD). But I am not sure that I can do that from my repository, now that it has been rebased?

raphaelcoeffic commented:
> > Regarding the issue with `VCENTERED`, where do you add it? It seems to fail compilation.
> 
> lua/api_general.cpp line 1870. I would move it up in a COLORLCD block, e.g. line1864 under FONT(BOLD). But I am not sure that I can do that from my repository, now that it has been rebased?

Of course you can, but you need to "re-checkout" the code from remote:
```
git fetch origin
git reset --hard 2.4 origin/2.4
```

But beware: this will wipe any local changes.

jfrickmann commented:
OK no problem. All local changes have already been pushed before this morning. I may not get around to it before tomorrow, as I am at work today, and we have a combined 25 yrs wedding anniversary + son's 20th birthday to celebrate tonight :)
And then I will read up on that Git story - I really need to do that!

jfrickmann commented:
I just edited the file here on GitHub, and now all checks have passed.

raphaelcoeffic commented:
> I just edited the file here on GitHub, and now all checks have passed. Still need to do the rebasing.

Rebasing is done, but you need to checkout the rebased code, that’s it.

wimalopaan commented:
Well, days ago I started to locally merge this with the main remote. Since yesterday(!) I get simulator24 and simulator25 and companion24 and companion25 alltogether. Looks like there is now something wrong, althoug no merge conflicts. 
And simulator24 requires 2.5V0037, thats strange.

Maybe I did something wrong here due to my limited git skills ... 
Any help?

JimB40 commented:
Flashed FW from artifact and put test widget on SD. Well done.

Still I wonder where these first 3 options come from?
![image](https://user-images.githubusercontent.com/46420768/126457625-3b4f5e71-20af-4ca6-91da-9f9bc9945215.jpeg)

jfrickmann commented:
OK, I have now read up on that Git-Fu stuff, and cleaned up my branches! I have split and rebased the commits by cherry picking them onto two fresh branches off main. I did some conflict resolution using meld. So this PR is back to the original topic, and I will create a new PR for events to the widgets.

jfrickmann commented:
Please see #489 for Lua widget events.
